flax/nnx/variablelib.py

```diff
@@ -966,3 +964,51 @@ def split_flat_state(
       )
 
   return flat_states
+
+
+
+###################################################
+### Variable type/class <-> string name mapping ###
+###################################################
+# Assumption: the mapping is 1-1 and unique.
+
+VariableTypeCache: dict[str, tp.Type[Variable[tp.Any]]] = {}
+
+
+def named_variable(name: str) -> tp.Type[Variable[tp.Any]]:
+  """Given a Linen-style collection name, get or create its NNX Variable class."""
+  if name not in VariableTypeCache:
+    VariableTypeCache[name] = type(name, (Variable,), {})
+  return VariableTypeCache[name]
+
+
+def variable_name(typ: tp.Type[Variable[tp.Any]]) -> str:
```

Comment:
Name suggestions: `variable_type_from_name` and `variable_name_from_type`. A bit more verbose but current names are a bit ambiguous.